### DISCUSSION TITLE
[PGSQL] Should we use a singleton pattern for DAO class in postgres-mailbox?

### DISCUSSION THREAD

**Discussion by vttranlina** [linagora/james-project#5002] [Announcements]
Currently, for each mailbox request, 
James will create a new DAO for each Mailbox session. 
https://github.com/apache/james-project/blob/a27db56f89252e8686b54a388a92b0631274342c/mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresMailboxSessionMapperFactory.java#L49

The reason for that is when we use RLS for the mailbox table (get Connection for each domain)

We can optimize it when we disable RLS, we can create a singleton DAO 
I can imagine, it will be quite verbose in the source code. 
WDYT?

**Comment by Arsnael** [comment]
We could groom this and create an investigation task around it I think before engulfing hours into dev.

But thanks for opening the discussion, it's needed I think :)

**Comment by quantranhong1999** [comment]
> I can imagine, it will be quite verbose in the source code.

  Very first thought in my head:

```java
  // create a Named bean e.g for PostgresSubscriptionMapper, and mark it as Singleton

  // create a PostgresSubscriptionMapperFactory
  public PostgresSubscriptionMapperFactory(@Named(NO_RLS) SubscriptionMapper subscriptionMapper, PostgresConfiguration postgresConfiguration) {
  }

  public SubscriptionMapper createSubscriptionMapper(MailboxSession session) {
        if (postgresConfiguration.rowLevelSecurityEnabled()) {
            create a new instance
        } else {
            return the singleton bean
        }
  }

   @Override
    public SubscriptionMapper createSubscriptionMapper(MailboxSession session) {
        return postgresSubscriptionMapperFactory.createSubscriptionMapper(session);
    }
```

**Reply by chibenwa** [reply]
Or 2 different factories depending on rls value

**Reply by vttranlina** [reply]
We will have 1 interface + 2 factories impl for each DAO?
